exonum/src/events/noise/tests.rs

```diff
@@ -84,3 +99,253 @@ fn test_converted_keys_handshake() {
+        let (public_key, secret_key) = gen_keypair_from_seed(&Seed::new([0; 32]));
+        HandshakeParams {
+            max_message_len: MAX_MESSAGE_LEN as u32,
+            public_key,
+            secret_key,
+        }
+    }
+}
+
+#[test]
+#[should_panic(expected = "WrongMessageLength(0)")]
+fn test_noise_handshake_errors_ee_empty() {
+    let addr: SocketAddr = "127.0.0.1:45003".parse().unwrap();
+    let step = HandshakeStep::EphemeralKeyExchange(1, EMPTY_MESSAGE);
+    let (sender_err, listener_err) = wait_for_handshake_result(&addr, step);
+
+    assert!(sender_err.is_err());
+    listener_err.unwrap()
+}
+
+#[test]
+#[should_panic(expected = "WrongMessageLength(0)")]
+fn test_noise_handshake_errors_es_empty() {
+    let addr: SocketAddr = "127.0.0.1:45004".parse().unwrap();
+    let step = HandshakeStep::StaticKeyExchange(2, EMPTY_MESSAGE);
+    let (sender_err, listener_err) = wait_for_handshake_result(&addr, step);
+
+    assert!(sender_err.is_err());
+    listener_err.unwrap()
+}
+
+#[test]
+#[should_panic(expected = "HandshakeNotFinished")]
+fn test_noise_handshake_errors_ee_standard() {
+    let addr: SocketAddr = "127.0.0.1:45005".parse().unwrap();
+    let step = HandshakeStep::EphemeralKeyExchange(1, STANDARD_MESSAGE);
+    let (sender_err, listener_err) = wait_for_handshake_result(&addr, step);
+
+    assert!(listener_err.is_err());
+    sender_err.unwrap()
+}
+
+#[test]
+#[should_panic(expected = "HandshakeNotFinished")]
+fn test_noise_handshake_errors_es_standard() {
+    let addr: SocketAddr = "127.0.0.1:45006".parse().unwrap();
+    let step = HandshakeStep::StaticKeyExchange(2, STANDARD_MESSAGE);
+    let (sender_err, listener_err) = wait_for_handshake_result(&addr, step);
+
+    assert!(listener_err.is_err());
+    sender_err.unwrap()
+}
+
+#[test]
+#[should_panic(expected = "Message size exceeds max handshake message size")]
+fn test_noise_handshake_errors_ee_large() {
+    let addr: SocketAddr = "127.0.0.1:45007".parse().unwrap();
+    let step = HandshakeStep::EphemeralKeyExchange(1, LARGE_MESSAGE);
+    let (sender_err, listener_err) = wait_for_handshake_result(&addr, step);
+
+    assert!(listener_err.is_err());
+    sender_err.unwrap()
+}
+
+#[test]
+#[should_panic(expected = "Message size exceeds max handshake message size")]
+fn test_noise_handshake_errors_se_large() {
+    let addr: SocketAddr = "127.0.0.1:45008".parse().unwrap();
+    let step = HandshakeStep::StaticKeyExchange(2, LARGE_MESSAGE);
+    let (sender_err, listener_err) = wait_for_handshake_result(&addr, step);
+
+    assert!(listener_err.is_err());
+    sender_err.unwrap()
+}
+
+// We need check result from both: sender and responder.
+fn wait_for_handshake_result(
+    addr: &SocketAddr,
+    step: HandshakeStep,
+) -> (IoResult<()>, IoResult<()>) {
+    let addr2 = addr.clone();
+
+    let (tx, rx) = mpsc::channel(0);
+    let (err_tx, err_rx) = mpsc::channel::<IoResult<()>>(0);
+    let rx = add_timeout_millis(rx, 500);
```

Comment:
I'm not sure why this construction is required, as well as the "only to connect" handshake. It seems that the problem could be solved by `sleep`ing after starting the listener (there may exist other solutions, such as channel-based signaling).